File: 0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null) return head;
        ListNode odd_current = head;
        ListNode even_current = head.next;
        ListNode even_head = head.next;
        while (even_current != null && even_current.next != null) {
            odd_current.next = even_current.next;
            odd_current = odd_current.next;
            even_current.next = odd_current.next;
            even_current = even_current.next;
        }
        odd_current.next = even_head;
        return head;
    }
}

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == nullptr || head->next == nullptr) return head;
        vector<int> arr;
        ListNode* temp = head;
        while (temp != nullptr && temp->next != nullptr) {
            arr.push_back(temp->val);
            temp = (temp->next)->next;
        }
        if (temp) arr.push_back(temp->val);
        temp = head->next;
        while (temp != nullptr && temp->next != nullptr) {
            arr.push_back(temp->val);
            temp = (temp->next)->next;
        }
        if (temp) arr.push_back(temp->val);
        temp = head;
        int idx = 0;
        while (temp != nullptr && idx < arr.size()) {
            temp->val = arr[idx++];
            temp = temp->next;
        }
        return head;
    }
};